subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileHierarchySet.java

```diff
@@ -82,6 +87,34 @@ public FileHierarchySet invalidate(String absolutePath) {
             .orElse(empty());
     }
 
+    @Override
+    public FileHierarchySet empty() {
+        return empty(caseSensitivity);
+    }
+
+    @Override
+    public void visitKnownDirectories(Consumer<File> directoryVisitor) {
+        Set<File> visited = new HashSet<>();
+        rootNode.accept((parentToRoot, node) -> {
+            File file = new File(new File("/"), parentToRoot);
+            node.getSnapshot().ifPresent(snapshot -> {
+                if (snapshot.getType() == FileType.Directory) {
+                    if (visited.add(file)) {
+                        directoryVisitor.accept(file);
+                    }
+                }
+            });
+            File parentFile = file;
+            while (true) {
+                parentFile = parentFile.getParentFile();
```

Comment:
Why are you going through all the parent files here? Would that cause to watch `/`?

Reply:
The JDK watcher on macOS doesn't support watching subdirectories, so we need to watch each individual directory separately. If the whole project disappears, or `/Users` disappears, we want to know.

Reply:
And yes, it means we are watching `/` as well, though it's very unlikely that there will be any activity there. Again, we are watching just the directory itself, not the whole hierarchy.